File: Medium/0494-target-sum/0494-target-sum.cpp
class Solution {
public:
    int solve(vector<int> &nums, int ind, int sum, int target, unordered_map<string, int> &memo) {
        string key = to_string(ind) + "," + to_string(sum);
        if (memo.find(key) != memo.end()) {
            return memo[key];
        }
        if (ind == nums.size()) {
            return sum == target ? 1 : 0;
        }
        memo[key] = solve(nums, ind + 1, sum + nums[ind], target, memo) + solve(nums, ind + 1, sum - nums[ind], target, memo);
        return memo[key];
    }
    
    int findTargetSumWays(vector<int>& nums, int target) {
        unordered_map<string, int> memo;
        return solve(nums, 0, 0, target, memo);
    }
};

File: Medium/0494-target-sum/0494-target-sum.py
class Solution:
    def findTargetSumWays(self, nums: List[int], target: int) -> int:
        @cache
        def solve(ind: int, sum: int) -> int:
            if ind == len(nums):
                return 1 if sum == target else 0
            return solve(ind + 1, sum + nums[ind]) + solve(ind + 1, sum - nums[ind])
        
        return solve(0, 0)